crypto/params.c

```diff
@@ -1208,6 +1230,12 @@ int OSSL_PARAM_get_double(const OSSL_PARAM *p, double *val)
     return 0;
 }
 
+# define D_POW_31 ((double) (((uint32_t) 1) << 31))
+static double d_pow_31 = D_POW_31;
+static double d_pow_32 = 2.0 * D_POW_31;
+static double d_pow_63 = 2.0 * D_POW_31 * D_POW_31;
+static double d_pow_64 = 4.0 * D_POW_31 * D_POW_31;
```

Comment:
These should be declared const.

Better would be putting them inside the function as non-static const declarations.

Reply:
They're file-scoped constants, adding the missing `const` qualifiers. I think this is now clear enough.

Reply:
You missed the point: there are only used by one function.

There is no reason for them to be file scoped, they should be function scoped.

Reply:
> You missed the point: there are only used by one function.
> 
> There is no reason for them to be file scoped, they should be function scoped.

When, in another PR, I create a function scoped static, Tim rather insisted it should be file-scoped.
I don't much care, perhaps you are willing to not care either? Otherwise, fight it out with Tim???

Reply:
They don't need to be static and IMO they shouldn't be.  Just declare them `const double`.

Reply:
My view was if they need to be static they should be file scoped rather than function scoped.
If they don't need to be static they should just be in the function. 
But that is a "style" issue more than anything else.

When we see a static in a function scope we are generally seeing an initialiser pattern where a value is set inside the function and needs to be seen outside the function consistently.

So I'm happy with either static file scope or non-static inside a function - but static in a function requires having to go and check to see if there are initialisation/threading issues.

@paulidale does that align with your thinking here?

Reply:
I am looking to initialise these at compile-time. The initialisers of static function-scoped variables are evaluated at compile-time, just like file-scoped statics, the only difference is the scope. That's all.

It is only **assignments** to function or file-scoped statics that need some scrutiny for races or other issues.
So I am not a fan of Tim's view on function scoped **const** statics, they're just fine.

Thus my proposal is function-scoped constants, initialised at compile time (stored in BSS) and never changed, but avoiding unnecessary file scope. I hope you're both OK with that.

Reply:
So what's the resolution? Can I (please) use function scope **static const** values? They're not runtime assignments, they're compile-time initialisations.

Reply:
We use function scoped statics all the time, so I wouldn't preclude them so widely.

Stop trying to out think the compiler.  I know of no compiler that will do this calculation at run time.  There is no performance benefit from making these static and having the static keyword means slightly more cognitive load to understand the code -- why is `static` there?

Just declare them const double in function scope:
```c
    const double d_pow_31 = D_POW_31;
    const double d_pow_32 = 2.0 * D_POW_31;
    const double d_pow_63 = 2.0 * D_POW_31 * D_POW_31;
    const double d_pow_64 = 4.0 * D_POW_31 * D_POW_31;
```